packages/gatsby-source-filesystem/src/gatsby-node.js

```diff
@@ -1,61 +1,124 @@
+  let pathQueue = []
+  const flushPathQueue = () => {
+    let queue = pathQueue.slice()
+    pathQueue = null
+    return Promise.all(queue.map(createAndProcessNode))
+  }
+
+  const createAndProcessNode = path => {
+    const fileNodePromise = createFileNode(
+      path,
+      createNodeId,
+      pluginOptions
+    ).then(fileNode => {
+      createNode(fileNode)
+      return null
+    })
+    return fileNodePromise
+  }
+
+  const fsMachine = Machine(
+    {
+      id: `fs`,
+      type: `parallel`,
+      states: {
+        BOOTSTRAP: {
+          initial: `BOOTSTRAPPING`,
+          states: {
+            BOOTSTRAPPING: {
+              on: {
+                BOOTSTRAP_FINISHED: `BOOTSTRAPPED`,
+              },
+              exit: `flushPathQueue`,
             },
-          },
-          CHOKIDAR_WATCHING: {
-            on: {
-              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
+            BOOTSTRAPPED: {
+              type: `final`,
             },
           },
-          CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED: {
-            on: {
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+        },
+        CHOKIDAR: {
+          initial: `NOT_READY`,
+          states: {
+            NOT_READY: {
+              on: {
+                CHOKIDAR_READY: `READY`,
+                CHOKIDAR_ADD: { actions: `queueNodeProcessing` },
+              },
             },
+            READY: {
+              on: {
+                CHOKIDAR_ADD: { actions: `createAndProcessNode` },
+              },
+            },
+          },
+          // TODO: those two were not restricted to READY state, but maybe they should? or even maybe it should queue in NOT_READY?
+          on: {
+            CHOKIDAR_CHANGE: { actions: `createAndProcessNode` },
+            CHOKIDAR_UNLINK: { actions: `deleteNode` },
           },
         },
       },
     },
-  })
-
-exports.sourceNodes = (
-  { actions, getNode, createNodeId, reporter, emitter },
-  pluginOptions
-) => {
-  const { createNode, createTypes, deleteNode } = actions
+    {
+      actions: {
+        createAndProcessNode(_, { pathType, path }, { state }) {
+          if (state.matches(`BOOTSTRAP.BOOTSTRAPPED`)) {
+            reporter.info(`added ${pathType} at ${path}`)
+          }
+          createAndProcessNode(path).catch(err => reporter.error(err))
+        },
+        deleteNode(_, { pathType, path }, { state }) {
+          if (state.matches(`BOOTSTRAP.BOOTSTRAPPED`)) {
+            reporter.info(`${pathType} deleted at ${path}`)
+          }
+          const node = getNode(createNodeId(path))
+          // It's possible the node was never created as sometimes tools will
+          // write and then immediately delete temporary files to the file system.
+          if (node) {
+            deleteNode({ node })
+          }
+        },
+        flushPathQueue(_, { resolve, reject }) {
```

Comment:
I spent embarrassingly long time looking where `resolve` / `reject` is coming from (initially I thought this is how `xstate` handles async action, before I found that `resolve`/`reject` is send in ``fsMachine.send({ type: `CHOKIDAR_READY`, resolve, reject })``

I do think we can rework it to be more inline with idiomatic xstate by:
 - creating new state in `CHOKIDAR` (sibling to `CHOKIDAR.NOT_READY`) - let's call it `FLUSHING` 
```diff
		  states: {
            NOT_READY: {
              on: {
-               CHOKIDAR_READY: `READY`,
+               CHOKIDAR_READY: `FLUSHING`,
                CHOKIDAR_ADD: { actions: `queueNodeProcessing` },
              },
            },
+           FLUSHING: {
+             invoke: {
+               src: () => {
+                 return flushPathQueue()
+               },
+               onDone: {
+                 target: `READY`,
+               },
+             },
+           },
            READY: {
              on: {
                CHOKIDAR_ADD: { actions: `createAndProcessNode` },
              },
            },
          },
```
see https://xstate.js.org/docs/guides/communication.html#invoking-services for docs about `invoke` property

- change the promise returned from `sourceNodes` a bit, let ``chokidar.on(`ready`)`` just send event to machine, and let's wait for machine to reach `CHOKIDAR.READY` to resolve that promise

Making it this way also shows that there is another potential problem (it's good thing) - what happens currently when `chokidar` emits event while flushing is in progress? Right now we would not queue work, but instead would process it immediately - making order of processing potentially incorrect. So I think while we are flushing, we should queue chokidar events and when flushing finishes check if queue is empty (then we can go to `CHOKIDAR.READY` state, and if it's not - flush again and repeat that until queue is empty)

Reply:
I like that 👍

One question though - `flushPathQueue` might reject and this would stay unhandled. I have no idea under what circumstances it could reject though? So not sure how this should be handled and for what to await for in `ready` listener to reject returned promise there.

Reply:
And again - maybe we could pursue fixing this in a followup PR? Or do you prefer keeping it as part of this one?

Reply:
Right, my idea here is certainly not fully fleshed out (just starting point). 

And also agree let's not scope creep this PR - let's make this PR 1 to 1 refactor (keeping current behaviour with all its flaws). All those edge cases that were uncovered can be incrementally handled in future pull requests

Reply:
Cool, I believe it's 1 to 1 right now if I haven't made any stupid mistake. I still don't know how to test this properly though - so I would appreciate you taking a look at this or giving me guidance.